MatrixKit/Controllers/MXKAuthenticationViewController.m

```diff
@@ -496,6 +378,97 @@ - (void)refreshSupportedAuthFlow
     }
 }
 
+#pragma mark - Privates
+
+- (void)updateRESTClient
+{
+    NSString *homeserverURL = _homeServerTextField.text;
+    
+    if (homeserverURL.length)
+    {
+        // Check change
+        if ([homeserverURL isEqualToString:mxRestClient.homeserver] == NO)
+        {
+            mxRestClient = [[MXRestClient alloc] initWithHomeServer:homeserverURL andOnUnrecognizedCertificateBlock:^BOOL(NSData *certificate) {
+                
+                __block BOOL isTrusted;
+                dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+                
+                NSString *title = [NSBundle mxk_localizedStringForKey:@"ssl_could_not_verify"];
+                NSString *homeserverURLStr = [NSString stringWithFormat:[NSBundle mxk_localizedStringForKey:@"ssl_homeserver_url"], homeserverURL];
+                NSString *fingerprint = [NSString stringWithFormat:[NSBundle mxk_localizedStringForKey:@"ssl_fingerprint_hash"], @"SHA256"];
+                NSString *certFingerprint = [certificate SHA256AsHexString];
+                
+                NSString *msg = [NSString stringWithFormat:@"%@\n\n%@\n\n%@\n\n%@\n\n%@\n\n%@", [NSBundle mxk_localizedStringForKey:@"ssl_cert_not_trust"], [NSBundle mxk_localizedStringForKey:@"ssl_cert_new_account_expl"], homeserverURLStr, fingerprint, certFingerprint, [NSBundle mxk_localizedStringForKey:@"ssl_only_accept"]];
+                
+                alert = [[MXKAlert alloc] initWithTitle:title message:msg style:MXKAlertStyleAlert];
+                alert.cancelButtonIndex = [alert addActionWithTitle:[NSBundle mxk_localizedStringForKey:@"cancel"] style:MXKAlertActionStyleDefault handler:^(MXKAlert *alert){
+                    
+                    isTrusted = NO;
+                    dispatch_semaphore_signal(semaphore);
+                    
+                }];
+                [alert addActionWithTitle:[NSBundle mxk_localizedStringForKey:@"ssl_trust"] style:MXKAlertActionStyleDefault handler:^(MXKAlert *alert){
+                    
+                    isTrusted = YES;
+                    dispatch_semaphore_signal(semaphore);
+                    
+                }];
+                
+                dispatch_async(dispatch_get_main_queue(), ^{
+                    [alert showInViewController:self];
+                });
+                
+                dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
```

Comment:
Mmm, semaphores. Other options to this would be passing a block to the block that it could run when it's made up its mind, or just having a method it runs (which is how the layer below works). That or look into a promises library for Obj-C. This would avoid blocking a thread for the duration the dialog is displayed.